File: spoj/spoj_toe2_tic_tac_toe_ii.py
#
# Author: eloyhz
# Date: Sep/21/2020
#
# SPOJ TOE2 - Tic-Tac-Toe ( II )
# https://www.spoj.com/problems/TOE2/
# 

def is_valid(board):
    r = [0] * 3
    c = [0] * 3
    d = [0] * 2
    cx, co = 0, 0
    for row in board:
        cx += row.count('X')
        co += row.count('O')
#    print(f'{cx = }, {co = }')
    if cx < co or cx > co + 1:
        return False
    dot = False
    for i in range(3):
        for j in range(3):
            if board[i][j] == '.':
                dot = True
                continue
            if i == j:
                d[0] += (1 if board[i][j] == 'X' else -1)
                d[1] += (1 if board[3 - i - 1][i] == 'X' else -1)
            r[i] += (1 if board[i][j] == 'X' else -1)
    for j in range(3):
        for i in range(3):
            if board[i][j] == '.':
                continue
            c[j] += (1 if board[i][j] == 'X' else -1)
    cx_wins = 0
    co_wins = 0
    for x in r:
        if x == 3:
            cx_wins += 1
        elif x == -3:
            co_wins += 1
    for x in c:
        if x == 3:
            cx_wins += 1
        elif x == -3:
            co_wins += 1
    for x in d:
        if x == 3:
            cx_wins += 1
        elif x == -3:
            co_wins += 1
#    for b in board: print(b)
#    print(f'{r = }, {c = }, {d = }, {count = }')
    if cx_wins == co_wins == 0 and dot:
        return False
    if co_wins > 0 and cx > co:
        return False
    if cx_wins > 0 and co_wins > 0:
        return False
    return True
    # return count <= 1


if __name__ == '__main__':
    while True:
        line = input()
        if line == "end":
            break
        board = [line[:3], line[3:6], line[6:]]
#        for b in board: print(b)
        print('valid' if is_valid(board) else 'invalid')
        

File: spoj/spoj_toe2_tic_tac_toe_ii.cpp
//
// Author: eloyhz
// Date: Sep/22/2020
//
// SPOJ TOE2 - Tic-Tac-Toe ( II )
// https://www.spoj.com/problems/TOE2/
//

#include <bits/stdc++.h>

using namespace std;
typedef vector<string> vs;

bool is_win(vs& b)
{
    return b[0][0] == b[0][1] && b[0][0] == b[0][2] && b[0][0] != '.' ||
            b[1][0] == b[1][1] && b[1][0] == b[1][2] && b[1][0] != '.' ||
            b[2][0] == b[2][1] && b[2][0] == b[2][2] && b[2][0] != '.' ||
            b[0][0] == b[1][0] && b[0][0] == b[2][0] && b[0][0] != '.' ||
            b[0][1] == b[1][1] && b[0][1] == b[2][1] && b[0][1] != '.' ||
            b[0][2] == b[1][2] && b[0][2] == b[2][2] && b[0][2] != '.' ||
            b[0][0] == b[1][1] && b[0][0] == b[2][2] && b[0][0] != '.' ||
            b[2][0] == b[1][1] && b[2][0] == b[0][2] && b[2][0] != '.';
}

bool is_full(vs& b) 
{
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (b[i][j] == '.') {
                return false;
            }
        }
    }
    return true;
}

bool is_valid(vs b)
{
    queue<pair<vs, char>> q;
    vs u = {"...", "...", "..."};
    q.push({u, 'X'});
    while (!q.empty())  {
        auto x = q.front();
        q.pop();
        u = x.first;
        char turn = x.second;
        if (u == b) {
            return (is_win(b) || is_full(b)) ? true : false;
        }
        if (is_win(u))  {
            continue;
        }
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (u[i][j] == '.' && b[i][j] == turn)  {
                    auto v = u;
                    v[i][j] = turn;
                    q.push({v, turn == 'X' ? 'O' : 'X'});
                }
            }
        }
    }
    return false;
}

int main()
{
    int n;
    vs board;
    string s, t;

    while (true) {
        cin >> s;
        if (s == "end") {
            break;
        }
        board.clear();
        board.push_back(s.substr(0, 3));
        board.push_back(s.substr(3, 3));
        board.push_back(s.substr(6, 3));
        cout << (is_valid(board) ? "valid" : "invalid") << endl;
   }
   return 0;
} 


